test/gtest/uct/test_flush.cc

```diff
@@ -538,3 +539,265 @@ UCS_TEST_SKIP_COND_P(uct_flush_test, am_pending_flush_nb,
+
+        peer(uct_cancel_test *test, peer *peer) :
+            test(*test), m_peer(peer),
+            m_e(NULL), m_buf1k(NULL), m_buf8k(NULL), m_buf32(NULL) {}
+
+        void init() {
+            m_e = test.uct_test::create_entity(0, error_handler_cb);
+            test.m_entities.push_back(m_e);
+            m_buf1k = new mapped_buffer(1 * 1024, 0, *m_e);
+            m_buf8k = new mapped_buffer(8 * 1024, 0, *m_e);
+            m_buf32 = new mapped_buffer(32, 0, *m_e);
+            uct_iface_set_am_handler(m_e->iface(), 0, am_cb, &test,
+                                     UCT_CB_FLAG_ASYNC);
+        }
+
+        void cleanup() {
+            if (m_buf1k) {
+                delete m_buf1k;
+                delete m_buf8k;
+                delete m_buf32;
+                m_buf1k = NULL;
+            }
+        }
+
+        ~peer() {
+            cleanup();
+        }
+
+        void connect() {
+            m_e->connect(0, *m_peer->m_e, 0);
+            m_peer->m_e->connect(0, *m_e, 0);
+        }
+    };
+
+    peer s0;
+    peer s1;
+
+    virtual void init() {
+        uct_test::init();
+
+        s0.init();
+        check_skip_test_tl();
+        s1.init();
+
+        s0.connect();
+        flush();
+    }
+
+    virtual void cleanup() {
+        flush();
+
+        s0.cleanup();
+        s1.cleanup();
+
+        uct_test::cleanup();
+    }
+
+    static ucs_status_t
+    error_handler_cb(void *arg, uct_ep_h ep, ucs_status_t status) {
+        uct_cancel_test *test = reinterpret_cast<uct_cancel_test*>(arg);
+        return test->error_handler(ep, status);
+    }
+
+    static ucs_status_t am_cb(void *arg, void *data, size_t length, unsigned flags) {
+        uct_cancel_test *test = reinterpret_cast<uct_cancel_test*>(arg);
+        return test->am(data, length, flags);
+    }
+
+    ucs_status_t error_handler(uct_ep_h ep, ucs_status_t status) {
+        EXPECT_EQ(UCS_ERR_CANCELED, status);
+        return UCS_OK;
+    }
+
+    ucs_status_t am(void *data, size_t length, unsigned flags) {
+        return UCS_OK;
+    }
+
+    void check_skip_test_tl() {
+        const resource *r = dynamic_cast<const resource*>(GetParam());
+
+        if ((r->tl_name != "rc_mlx5") &&
+            (r->tl_name != "rc_verbs")) {
+            UCS_TEST_SKIP_R("not supported yet");
+        }
+
+        check_skip_test();
+    }
+
+public:
+    uct_cancel_test() : uct_test(), s0(this, &s1), s1(this, &s0) {}
+
+    ucs_status_t am_bcopy(peer *s) {
+        mapped_buffer &sendbuf = *s->m_buf32;
+        ssize_t packed_len;
+
+        packed_len = uct_ep_am_bcopy(s->m_e->ep(0), 0, mapped_buffer::pack,
+                                     (void*)&sendbuf, 0);
+        if (packed_len >= 0) {
+            EXPECT_EQ(sendbuf.length(), (size_t)packed_len);
+            return UCS_OK;
+        } else {
+            return (ucs_status_t)packed_len;
+        }
+    }
+
+    ucs_status_t am_zcopy(peer *s) {
+        mapped_buffer &sendbuf = *s->m_buf8k;
```

Comment:
how can we know that 8k is really supported by am_zcopy? test should be more generic and check zcopy max
same for other ops